packages/ckeditor5-clipboard/src/utils/viewtoplaintext.ts

```diff
@@ -21,41 +21,61 @@ const smallPaddingElements = [ 'figcaption', 'li' ];
  * @returns Plain text representation of `viewItem`.
  */
 export default function viewToPlainText( viewItem: ViewItem | ViewDocumentFragment ): string {
+	if ( viewItem.is( '$text' ) || viewItem.is( '$textProxy' ) ) {
+		return viewItem.data;
+	}
+
+	if ( viewItem.is( 'element', 'img' ) && viewItem.hasAttribute( 'alt' ) ) {
+		return viewItem.getAttribute( 'alt' )!;
+	}
+
+	if ( viewItem.is( 'element', 'br' ) ) {
+		return '\n'; // Convert soft breaks to single line break (#8045).
+	}
+
+	/**
+	 * Item is a document fragment, attribute element or container element. It doesn't
+	 * have it's own text value, so we need to convert its children elements.
+	 */
+
 	let text = '';
+	let prev: ViewElement | null = null;
 
-	if ( viewItem.is( '$text' ) || viewItem.is( '$textProxy' ) ) {
-		// If item is `Text` or `TextProxy` simple take its text data.
-		text = viewItem.data;
-	} else if ( viewItem.is( 'element', 'img' ) && viewItem.hasAttribute( 'alt' ) ) {
-		// Special case for images - use alt attribute if it is provided.
-		text = viewItem.getAttribute( 'alt' )!;
-	} else if ( viewItem.is( 'element', 'br' ) ) {
-		// A soft break should be converted into a single line break (#8045).
-		text = '\n';
-	} else {
-		// Other elements are document fragments, attribute elements or container elements.
-		// They don't have their own text value, so convert their children.
-		let prev = null;
-
-		for ( const child of ( viewItem as ViewElement | ViewDocumentFragment ).getChildren() ) {
-			const childText = viewToPlainText( child );
-
-			// Separate container element children with one or more new-line characters.
-			if ( prev && ( prev.is( 'containerElement' ) || child.is( 'containerElement' ) ) ) {
-				if (
-					smallPaddingElements.includes( ( prev as ViewElement ).name ) ||
-					smallPaddingElements.includes( ( child as ViewElement ).name )
-				) {
-					text += '\n';
-				} else {
-					text += '\n\n';
-				}
-			}
-
-			text += childText;
-			prev = child;
-		}
+	for ( const child of ( viewItem as ViewElement | ViewDocumentFragment ).getChildren() ) {
+		text += newLinePadding( child as ViewElement, prev ) + viewToPlainText( child );
+		prev = child as ViewElement;
 	}
 
 	return text;
 }
+
+/**
+ * Returns new line padding to prefix the given elements with.
+ */
+function newLinePadding(
+	element: ViewElement,
+	previous: ViewElement | null
+): string {
+	if ( !previous ) {
+		// Don't add padding to first elements in a level.
+		return '';
+	}
+
+	if ( element.is( 'element', 'li' ) && !element.isEmpty && !element.getChild( 0 )!.is( '$text' ) ) {
+		// Separate document list items with empty lines.
+		return '\n\n';
+	}
```

Comment:
If I understand this correctly, it's supposed to detect list items with blocks inside. This would fail on such a structure:

```html
<ul>
  <li>UL List item 1</li>
  <li><strong>UL List item 2</strong></li>
</ul>
```
The second LI would be detected as DocumentList but it's not in legacy lists.

Also, the code above would not detect properly simple lists in document lists.

Reply:
I pushed fixes to:

* properly detect if element is a document list (https://github.com/ckeditor/ckeditor5/pull/15042/commits/8c16e9eecef64094a869060ae27c8a9472bac3b6),
* avoid breaking simple lists when pasting paragraphs or plain text lists (https://github.com/ckeditor/ckeditor5/pull/15042/commits/906ee8060383c1a3723a4043de78d9ab1df1fa39).